### DISCUSSION TITLE
How to use the pb to impl a grpc client

### DISCUSSION THREAD

**Discussion by caibirdme** [vectordotdev/vector#12456] [Q&A] [answered]
https://github.com/vectordotdev/vector/blob/f1f1be7c3e9862b38a2a411e3fe96bef61dfb83e/lib/vector-core/proto/event.proto#L35

I want to implement a grpc client to send logs directly to vector aggregator, because the performance of grpc is much better than http.
I found protobuf definitions here, but don't know how to use it. Especially how to set the Key of the map, what does the key mean? Are there any required keys or any conventions? Could you please give some Pseudo code to explain that?

**Comment by jszwedko** [comment, marked as answer]
Hi @caibirdme ! The map key there is just the field name. E.g. if you have a log event that looks like:

```json
{
  "message": "hello world"
}
```

The key would be "message" and the value would be "hello world".

**Reply by jszwedko** [reply]
There are no required keys for ingest. Some sinks require some keys (like `host` or `timestamp`).

**Reply by jszwedko** [reply]
Also `message` is a common key used by sinks.
